toolkit/popup/src/main/java/com/arcgismaps/toolkit/popup/internal/ui/fileviewer/ViewableFile.kt

```diff
@@ -0,0 +1,146 @@
+            else -> ViewableFileType.Other
+        }
+    }
+
+    override fun ViewableFileType.write(parcel: Parcel, flags: Int) {
+        parcel.writeInt(
+            when (this) {
+                ViewableFileType.Image -> 0
+                ViewableFileType.Video -> 1
+                ViewableFileType.Other -> 2
+            }
+        )
+    }
+}
+
+internal sealed class ViewableFileType {
+    data object Image : ViewableFileType()
+    data object Video : ViewableFileType()
+    data object Other : ViewableFileType()
+}
+
+internal fun PopupAttachmentType.toViewableFileType(): ViewableFileType = when (this) {
+    PopupAttachmentType.Image -> ViewableFileType.Image
+    PopupAttachmentType.Video -> ViewableFileType.Video
+    PopupAttachmentType.Document,
+    PopupAttachmentType.Other -> ViewableFileType.Other
+}
+
+/**
+ * Saves the file to the device.
+ */
+internal suspend fun ViewableFile.saveToDevice(context: Context): Result<Unit> = withContext(Dispatchers.IO) {
+    runCatching {
+        val sourceFile = File(path).takeIf { it.exists() }
+            ?: throw FileNotFoundException("File not found: $path")
+
+        // define the file values
+        val fileValues = ContentValues().apply {
+            put(MediaStore.Images.Media.DISPLAY_NAME, name)
+            put(MediaStore.Images.Media.MIME_TYPE, contentType)
+        }
+
+        @SuppressLint("InlinedApi")
+        val uri = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q)
+            MediaStore.VOLUME_EXTERNAL_PRIMARY
+        else
+            MediaStore.VOLUME_EXTERNAL
+
+        val contentCollection = when (type) {
+            ViewableFileType.Video -> MediaStore.Video.Media.getContentUri(uri)
+            ViewableFileType.Image -> MediaStore.Images.Media.getContentUri(uri)
+            else -> throw UnsupportedOperationException("Cannot save this file type")
+        }
+        val destinationUri = context.contentResolver.insert(contentCollection, fileValues)
+            ?: throw IOException("Failed to save file")
+
+        // copy file to destination
+        val sourceUri = Uri.fromFile(sourceFile)
+        context.contentResolver?.openInputStream(sourceUri)?.use { inputStream ->
+            context.contentResolver.openOutputStream(destinationUri)?.use { outputStream ->
+                val buffer = ByteArray(1024)
+                var length: Int
+                while (inputStream.read(buffer).also { length = it } > 0) {
+                    outputStream.write(buffer, 0, length)
+                }
+            }
+        } ?: throw IOException("Failed to save file")
+    }
+}
+
+/**
+ * Shares the file using Android's share sheet.
+ */
+internal suspend fun ViewableFile.share(context: Context) = withContext(Dispatchers.IO) {
+    val file = File(path)
+
+    val uri = FileProvider.getUriForFile(
+        context.applicationContext,
+        "${context.applicationContext.applicationInfo.packageName}.fileprovider",
```

Comment:
If changes in the manifest are made to make the authority more unique, this will need to change as well.
```suggestion
        "${context.applicationContext.applicationInfo.packageName}.arcgis.popup.fileprovider",
```